assets/src/edit-story/components/canvas/utils/useSnapping.js

```diff
@@ -14,34 +14,100 @@
  * limitations under the License.
  */
 
+/**
+ * External dependencies
+ */
+import { useCallback } from 'react';
+
 /**
  * Internal dependencies
  */
 import { useGlobalIsKeyPressed } from '../../keyboard';
 import useCanvas from '../useCanvas';
 
 function useSnapping({ canSnap, otherNodes }) {
-  const { canvasWidth, canvasHeight } = useCanvas(
+  const {
+    canvasWidth,
+    canvasHeight,
+    pageContainer,
+    canvasContainer,
+    designSpaceGuideline,
+    showSafeZone,
+  } = useCanvas(
     ({
       state: {
         pageSize: { width: canvasWidth, height: canvasHeight },
+        pageContainer,
+        canvasContainer,
+        designSpaceGuideline,
+        showSafeZone,
       },
-    }) => ({ canvasWidth, canvasHeight })
+    }) => ({
+      canvasWidth,
+      canvasHeight,
+      pageContainer,
+      canvasContainer,
+      designSpaceGuideline,
+      showSafeZone,
+    })
   );
 
   // ⌘ key disables snapping
   const snapDisabled = useGlobalIsKeyPressed('meta');
   canSnap = canSnap && !snapDisabled;
+
+  const toggleDesignSpace = useCallback(
+    (visible) => {
+      designSpaceGuideline.style.visibility = visible ? 'visible' : 'hidden';
+    },
+    [designSpaceGuideline]
+  );
+  const handleSnap = useCallback(
+    ({ elements }) =>
+      // Show design space if we're snapping to any of its edges
+      toggleDesignSpace(
+        elements
+          .flat()
+          .some(
+            ({ center, element }) => element === designSpaceGuideline && !center
+          )
+      ),
+    [toggleDesignSpace, designSpaceGuideline]
+  );
+
+  if (!canvasContainer || !pageContainer) {
+    return {};
+  }
+
+  const canvasRect = canvasContainer.getBoundingClientRect();
+  const pageRect = pageContainer.getBoundingClientRect();
+
+  const offsetX = Math.ceil(pageRect.x - canvasRect.x);
+  const offsetY = Math.floor(pageRect.y - canvasRect.y);
+
+  const verticalGuidelines = canSnap
+    ? [offsetX, offsetX + canvasWidth / 2, offsetX + canvasWidth]
+    : [];
+
+  const horizontalGuidelines = canSnap
+    ? [offsetY, offsetY + canvasHeight / 2, offsetY + canvasHeight]
+    : [];
+
+  const elementGuidelines = canSnap
+    ? [...otherNodes, ...(showSafeZone ? [designSpaceGuideline] : [])]
```

Comment:
Good idea adding it to the elements list.

Reply:
One thing I'm wondering about is if the design space should snap even if the element is touching it from the outside only (while being fully outside of the design space), this kind of defeats the purpose of keeping the elements inside the space. However, perhaps it's OK for now as-is. WDYT?

Reply:
But how would we do that? Yes, we can not show the design space guideline (the pink box) if it touches from the outside, but moveable would still show the vertical snapping guideline - we can't add a guideline or snapping element and say "only snap if on the left side" or something like that. So, we'd be showing vertical snapping guidelines in weird places, where it seems like there's nothing to snap to.

Reply:
Maybe our good friend @daybrush can help us out here.

Reply:
Hmm, I guess we could do the calculation during transformation to understand if the element is outside of the design space, and if it is, disable the relevant snapping guidelines by passing in only `otherNodes`.

Anyway, this behavior is not specified by UX currently so I think it's OK to leave it for the future if it becomes relevant.

Reply:
Hm, yeah, that's true - we could disable that as a snapping element as soon as the selection left boundary is over the left border or the selection right boundary is over the right border. But that's pretty complex. I'll leave that for now.